backend/api/v1/user_group.go

```diff
@@ -0,0 +1,195 @@
+	email, err := common.GetUserGroupEmail(request.Group.Name)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	principalID, ok := ctx.Value(common.PrincipalIDContextKey).(int)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "principal ID not found")
+	}
+
+	patch := &store.UpdateUserGroupMessage{}
+
+	for _, path := range request.UpdateMask.Paths {
+		switch path {
+		case "title":
+			patch.Title = &request.Group.Title
+		case "description":
+			patch.Description = &request.Group.Description
+		case "members":
+			payload, err := convertToGroupPayload(request.Group)
+			if err != nil {
+				return nil, status.Errorf(codes.InvalidArgument, err.Error())
+			}
+			patch.Payload = payload
+		default:
+			return nil, status.Errorf(codes.InvalidArgument, `unsupport update_mask "%s"`, path)
+		}
+	}
+
+	groupMessage, err := s.store.UpdateUserGroup(ctx, email, patch, principalID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	return convertToV1Group(groupMessage), nil
+}
+
+// DeleteUserGroup deletes a group.
+func (s *UserGroupService) DeleteUserGroup(ctx context.Context, request *v1pb.DeleteUserGroupRequest) (*emptypb.Empty, error) {
+	email, err := common.GetUserGroupEmail(request.Name)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	if err := s.store.DeleteUserGroup(ctx, email); err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func convertToGroupPayload(group *v1pb.UserGroup) (*storepb.UserGroupPayload, error) {
+	payload := &storepb.UserGroupPayload{}
+	for _, member := range group.Members {
+		m := &storepb.UserGroupMember{
+			Member: member.Member,
+		}
+		switch member.Role {
+		case v1pb.UserGroupMember_MEMBER:
+			m.Role = storepb.UserGroupMember_MEMBER
+		case v1pb.UserGroupMember_OWNER:
+			m.Role = storepb.UserGroupMember_OWNER
+		default:
+			return nil, status.Errorf(codes.InvalidArgument, "unsupport group member role %v", member.Role)
+		}
+		payload.Members = append(payload.Members, m)
+	}
+	return payload, nil
+}
+
+func convertToGroupMessage(group *v1pb.UserGroup) (*store.UserGroupMessage, error) {
+	email, err := common.GetUserGroupEmail(group.Name)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	groupMessage := &store.UserGroupMessage{
+		Email:       email,
+		Title:       group.Title,
+		Description: group.Description,
+		Payload:     &storepb.UserGroupPayload{},
+	}
+
+	payload, err := convertToGroupPayload(group)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+	groupMessage.Payload = payload
+	return groupMessage, nil
+}
+
+func convertToV1Group(groupMessage *store.UserGroupMessage) *v1pb.UserGroup {
+	group := &v1pb.UserGroup{
+		Name:        fmt.Sprintf("%s%s", common.UserGroupPrefix, groupMessage.Email),
```

Comment:
use `common.FormatUserEmail`

Reply:
It's not user email, unless we implement another function with group prefix

Reply:
You are right. I misread.